Loading dataset(Wine Quality dataset)

In [2]:
import numpy as np

data=np.loadtxt("../Data/winequality-red.csv", delimiter=";", skiprows=1)


Inspecting the data

In [3]:
#Inspecting the number of dimensions and shape of the data

print(data.shape)
print(data.ndim)

#Inspecting the data types of the columns
print(data.dtype)

(1599, 12)
2
float64


Retrieving first wine and last wine in the dataset

In [4]:
#first wine
print(data[0,:])

#last wine
print(data[-1])

[ 7.4     0.7     0.      1.9     0.076  11.     34.      0.9978  3.51
  0.56    9.4     5.    ]
[ 6.       0.31     0.47     3.6      0.067   18.      42.       0.99549
  3.39     0.66    11.       6.     ]


In [5]:
#Retrieving the first 5 rows of the data
np.set_printoptions(suppress="True")
data[:5,:]

array([[ 7.4   ,  0.7   ,  0.    ,  1.9   ,  0.076 , 11.    , 34.    ,
         0.9978,  3.51  ,  0.56  ,  9.4   ,  5.    ],
       [ 7.8   ,  0.88  ,  0.    ,  2.6   ,  0.098 , 25.    , 67.    ,
         0.9968,  3.2   ,  0.68  ,  9.8   ,  5.    ],
       [ 7.8   ,  0.76  ,  0.04  ,  2.3   ,  0.092 , 15.    , 54.    ,
         0.997 ,  3.26  ,  0.65  ,  9.8   ,  5.    ],
       [11.2   ,  0.28  ,  0.56  ,  1.9   ,  0.075 , 17.    , 60.    ,
         0.998 ,  3.16  ,  0.58  ,  9.8   ,  6.    ],
       [ 7.4   ,  0.7   ,  0.    ,  1.9   ,  0.076 , 11.    , 34.    ,
         0.9978,  3.51  ,  0.56  ,  9.4   ,  5.    ]])

In [6]:
#Extracting alcohol values for each wine
data[:,10]

array([ 9.4,  9.8,  9.8, ..., 11. , 10.2, 11. ], shape=(1599,))

In [7]:
#Extracting quality column
data[:,-1]

array([5., 5., 5., ..., 6., 5., 6.], shape=(1599,))

Extracting these columns:                                                          
1.fixed acidity
2.volatile acidity
3.citric acid
4.alcohol
5.quality

In [8]:
data[:, [0,1,2,-2,-1]]

array([[ 7.4  ,  0.7  ,  0.   ,  9.4  ,  5.   ],
       [ 7.8  ,  0.88 ,  0.   ,  9.8  ,  5.   ],
       [ 7.8  ,  0.76 ,  0.04 ,  9.8  ,  5.   ],
       ...,
       [ 6.3  ,  0.51 ,  0.13 , 11.   ,  6.   ],
       [ 5.9  ,  0.645,  0.12 , 10.2  ,  5.   ],
       [ 6.   ,  0.31 ,  0.47 , 11.   ,  6.   ]], shape=(1599, 5))

Extracting every 10th wine starting from the first

In [9]:
# extracted=np.arange(0,data.shape[0],10)
# data[extracted].tolist()

print(data[::10])

[[ 7.4   0.7   0.   ...  0.56  9.4   5.  ]
 [ 6.7   0.58  0.08 ...  0.54  9.2   5.  ]
 [ 8.9   0.22  0.48 ...  0.53  9.4   6.  ]
 ...
 [ 6.4   0.36  0.53 ...  0.93 12.4   6.  ]
 [ 7.4   0.35  0.33 ...  0.6  11.9   6.  ]
 [ 6.3   0.55  0.15 ...  0.82 11.6   6.  ]]


In [10]:
#Reversing dataset
print(data[::-1])

[[ 6.     0.31   0.47  ...  0.66  11.     6.   ]
 [ 5.9    0.645  0.12  ...  0.71  10.2    5.   ]
 [ 6.3    0.51   0.13  ...  0.75  11.     6.   ]
 ...
 [ 7.8    0.76   0.04  ...  0.65   9.8    5.   ]
 [ 7.8    0.88   0.    ...  0.68   9.8    5.   ]
 [ 7.4    0.7    0.    ...  0.56   9.4    5.   ]]


In [11]:
#Calculating the mean of alcohol
average=np.mean(data[:,-2])

print(average)

10.422983114446529


Alcohol Extremes

Finding Minimum alcohol, Maximum alcohol, Mean alcohol, Median alcohol and Standard deviation of alcohol

In [12]:
#Minimum Value
min=np.min(data[:,-2], axis=0)
print("Minimum Alcohol was:",min)

#Maximum Value
max=np.max(data[:,-2], axis=0)
print("Maximum Alcohol was:",max)

#The Median
median=np.median(data[:,-2], axis=0)
print("Median Alcohol was:",median)

#The Standard  deviation
std=np.std(data[:,-2], axis=0)
print("Standard Deviation of Alcohol was:",std)


Minimum Alcohol was: 8.4
Maximum Alcohol was: 14.9
Median Alcohol was: 10.2
Standard Deviation of Alcohol was: 1.0653343003437483


In [13]:
#Calculating the minimum, maximum, mean, median and standard deviation of wine quality
min_q=np.min(data[:,-1], axis=0)
print("Minimum Quality was:",min_q)

max_q=np.max(data[:,-1], axis=0)
print("Maximum Quality was:",max_q)

mean_q=np.mean(data[:,-1], axis=0)
print("Mean Quality was:",mean_q)

median_q=np.median(data[:,-1], axis=0)
print("Median Quality was:",median_q)

std_q=np.std(data[:,-1], axis=0)
print("Standard Deviation of Quality was:",std_q)


Minimum Quality was: 3.0
Maximum Quality was: 8.0
Mean Quality was: 5.6360225140712945
Median Quality was: 6.0
Standard Deviation of Quality was: 0.8073168769639513


Finding wine with the highest Fixed Acidity

In [15]:
least_alcohol=np.argmin(data[:,-2])
print("Index of wine with least alcohol content",least_alcohol)

print("wine with least alcoholic percentage:\n",data[least_alcohol])

Index of wine with least alcohol content 517
wine with least alcoholic percentage:
 [10.4     0.61    0.49    2.1     0.2     5.     16.      0.9994  3.16
  0.63    8.4     3.    ]


Highest Quality Wines

In [16]:

high_q_col=data[:,-1]
Highest_quality_wine=np.where(high_q_col==np.max(high_q_col))
print("Wine with highest quality wine:\n",data[Highest_quality_wine])


Wine with highest quality wine:
 [[ 7.9      0.35     0.46     3.6      0.078   15.      37.       0.9973
   3.35     0.86    12.8      8.     ]
 [10.3      0.32     0.45     6.4      0.073    5.      13.       0.9976
   3.23     0.82    12.6      8.     ]
 [ 5.6      0.85     0.05     1.4      0.045   12.      88.       0.9924
   3.56     0.82    12.9      8.     ]
 [12.6      0.31     0.72     2.2      0.072    6.      29.       0.9987
   2.88     0.82     9.8      8.     ]
 [11.3      0.62     0.67     5.2      0.086    6.      19.       0.9988
   3.22     0.69    13.4      8.     ]
 [ 9.4      0.3      0.56     2.8      0.08     6.      17.       0.9964
   3.15     0.92    11.7      8.     ]
 [10.7      0.35     0.53     2.6      0.07     5.      16.       0.9972
   3.15     0.65    11.       8.     ]
 [10.7      0.35     0.53     2.6      0.07     5.      16.       0.9972
   3.15     0.65    11.       8.     ]
 [ 5.       0.42     0.24     2.       0.06    19.      50.       0.991

Boolean Masking

Finding all wines with alcohol greater than 10%

In [17]:
#Finding alcohol with alcohol>10%

#converting our dataset into numpy array
data=np.array(data)
column_names=np.array(["fixed acidity","volatile acidity","citric acid","residual sugar","chlorides","free sulfur dioxide","total sulfur dioxide","density","pH","sulphates","alcohol","quality"])

alcohol_col=column_names=="alcohol"
alcohol_col_val= data[:,alcohol_col].flatten()>10

alcohol_values_greater=data[alcohol_col_val]
alcohol_values_greater


array([[ 7.5  ,  0.5  ,  0.36 , ...,  0.8  , 10.5  ,  5.   ],
       [ 7.5  ,  0.5  ,  0.36 , ...,  0.8  , 10.5  ,  5.   ],
       [ 8.5  ,  0.28 ,  0.56 , ...,  0.75 , 10.5  ,  7.   ],
       ...,
       [ 6.3  ,  0.51 ,  0.13 , ...,  0.75 , 11.   ,  6.   ],
       [ 5.9  ,  0.645,  0.12 , ...,  0.71 , 10.2  ,  5.   ],
       [ 6.   ,  0.31 ,  0.47 , ...,  0.66 , 11.   ,  6.   ]],
      shape=(852, 12))

Displaying wines with high fixed acidity

In [60]:
#finding wine with the highest fixed acidity
fixed_acidity=column_names=="fixed acidity"

max_acidity_wine = np.max(data[:, 0])
print("value Wine with the highest fixed acidity:", max_acidity_wine)

max=data[:,fixed_acidity]==max_acidity_wine
#Entire row of the wine with the highest fixed acidity
print("Wines with the highest fixed acidity:\n", data[max.flatten()])



value Wine with the highest fixed acidity: 15.9
Wines with the highest fixed acidity:
 [[15.9   0.36  0.65  7.5   0.1  22.   71.    1.    2.98  0.84 14.9   5.  ]]


Displaying rows with wine quality greater than 7

In [18]:
quality=column_names=="quality"

wine_quality=data[:,quality]
all_wines_quality=wine_quality.flatten()>=7

print(data[all_wines_quality])

[[ 7.3   0.65  0.   ...  0.47 10.    7.  ]
 [ 7.8   0.58  0.02 ...  0.57  9.5   7.  ]
 [ 8.5   0.28  0.56 ...  0.75 10.5   7.  ]
 ...
 [ 7.4   0.36  0.3  ...  0.7  11.4   8.  ]
 [ 7.    0.56  0.17 ...  0.68 10.55  7.  ]
 [ 6.7   0.32  0.44 ...  0.8  11.6   7.  ]]


Displaying wines with alcohol greater than 10 and quality greater or equal to 7

In [19]:
Both_wines=all_wines_quality & alcohol_col_val
print(data[Both_wines])

[[ 8.5    0.28   0.56  ...  0.75  10.5    7.   ]
 [ 8.     0.59   0.16  ...  0.92  10.5    7.   ]
 [ 5.4    0.835  0.08  ...  0.85  13.     7.   ]
 ...
 [ 7.4    0.36   0.3   ...  0.7   11.4    8.   ]
 [ 7.     0.56   0.17  ...  0.68  10.55   7.   ]
 [ 6.7    0.32   0.44  ...  0.8   11.6    7.   ]]


Displaying wines where pH is less thsn 3 OR greater than 12%

In [20]:
#changing the Scientific notation to the actual values
np.set_printoptions(suppress=True, precision=2)


#Displaying wines where pH<3 or alcohol>12%
Ph_col=column_names=="pH"

ph_row=data[:,Ph_col].flatten()

all_ph_row=ph_row<3

alcohol_than_12=column_names=="alcohol"

wines_alcohol_than_12=data[:,alcohol_than_12]
all_wines_alcohol_than_12=wines_alcohol_than_12.flatten()>12


ph_or_alcohol=all_wines_alcohol_than_12| all_ph_row
print(data[ph_or_alcohol])


[[ 4.6   0.52  0.15 ...  0.56 13.1   4.  ]
 [ 8.6   0.49  0.28 ...  1.95  9.9   6.  ]
 [ 8.6   0.49  0.28 ...  1.95  9.9   6.  ]
 ...
 [ 6.4   0.36  0.53 ...  0.93 12.4   6.  ]
 [ 6.    0.58  0.2  ...  0.67 12.5   6.  ]
 [ 7.2   0.66  0.33 ...  0.78 12.8   6.  ]]


 Total counts of wines with alcohol>10%                                           
 Total counts of wines with quality>=7


In [21]:
#Num of alcohols greater than 10%
total_alcohol_greater_10=np.count_nonzero(all_wines_alcohol_than_12)
print("Total count of all alcohol greater than 10:",total_alcohol_greater_10)

#Sum of quality greater or equal to 7
total_quality_greaterOrEqual_7=np.count_nonzero(all_wines_quality)
print("Total count of all quality greater or equal to 7:",total_quality_greaterOrEqual_7)



Total count of all alcohol greater than 10: 141
Total count of all quality greater or equal to 7: 217


Percentage of all wines with quality greater or equal to 7

In [22]:
percentage=total_quality_greaterOrEqual_7/1599*100
print(f"Percentage of all wines with quality greater or equal to 7 is {percentage:.2f}%")

Percentage of all wines with quality greater or equal to 7 is 13.57%


Finding the wines with strong alcohol content

In [23]:
#Average alcohol content
alcohol=column_names=="alcohol"
average_alcohol=np.mean(data[:,alcohol])
print(f"Average alcohol is: {average_alcohol:.2f}")

#Fing wines with potentially strong alcohol content
strong_alcohol_cont=data[:,alcohol].flatten()
strong_alcohol_cont=strong_alcohol_cont[strong_alcohol_cont>average_alcohol]
print(f"There are {np.count_nonzero(strong_alcohol_cont)} wines with higher alcohol content")

Average alcohol is: 10.42
There are 683 wines with higher alcohol content


Find the wines whose alcohol content is above average AND quality is at least 7.

In [24]:
alcohol_cols=column_names=="alcohol"
alcohol_values=data[:,alcohol_cols].flatten()

alcohol_beyond_avg=alcohol_values>average_alcohol

quality_cols=column_names=="quality"

quality_values=data[:, quality_cols].flatten()

quality_atLeast_7=quality_values>=7

combined_alcohol_and_quality=alcohol_beyond_avg&quality_atLeast_7
print("Number of alcohol above average and quality ",np.sum(combined_alcohol_and_quality))

Number of alcohol above average and quality  188


Finding wines where volatile acidity is less than 0.5, alcohol is greater than 10 and quality is greater or equal to 6

In [25]:
volatil_col=column_names=="volatile acidity"
volatil_data=data[:,volatil_col].flatten()<0.5

alcohol1_cols=column_names=="alcohol"
alcohol1_values=data[:,alcohol1_cols].flatten()>10

quality1_cols=column_names=="quality"
quality1_values=data[:,quality1_cols].flatten()>=6

combined_rows=volatil_data & alcohol1_values & quality1_values
print("Rows where volatile acidity<0.5, alcohol>10 and quality>=6\n\n",data[combined_rows])




Rows where volatile acidity<0.5, alcohol>10 and quality>=6

 [[ 8.5   0.28  0.56 ...  0.75 10.5   7.  ]
 [ 7.5   0.49  0.2  ...  0.9  10.5   6.  ]
 [ 6.3   0.3   0.48 ...  0.78 10.3   6.  ]
 ...
 [ 7.2   0.39  0.44 ...  0.84 11.5   6.  ]
 [ 7.5   0.31  0.41 ...  0.85 11.4   6.  ]
 [ 6.    0.31  0.47 ...  0.66 11.    6.  ]]


Finding Top 10 highest quality wines

In [26]:
quality1=column_names=="quality"
sort_quality=np.argsort(data[:,quality1].flatten())
print("Top 10 high quality wines:\n",data[sort_quality[::-1][:10]])



Top 10 high quality wines:
 [[ 7.4   0.36  0.3   1.8   0.07 17.   24.    0.99  3.24  0.7  11.4   8.  ]
 [ 7.8   0.57  0.09  2.3   0.07 34.   45.    0.99  3.46  0.74 12.7   8.  ]
 [ 5.    0.42  0.24  2.    0.06 19.   50.    0.99  3.72  0.74 14.    8.  ]
 [ 7.9   0.54  0.34  2.5   0.08  8.   17.    0.99  3.2   0.72 13.1   8.  ]
 [10.    0.26  0.54  1.9   0.08 42.   74.    0.99  2.98  0.63 11.8   8.  ]
 [ 7.2   0.38  0.31  2.    0.06 15.   29.    0.99  3.23  0.76 11.3   8.  ]
 [ 9.4   0.3   0.56  2.8   0.08  6.   17.    1.    3.15  0.92 11.7   8.  ]
 [10.7   0.35  0.53  2.6   0.07  5.   16.    1.    3.15  0.65 11.    8.  ]
 [10.7   0.35  0.53  2.6   0.07  5.   16.    1.    3.15  0.65 11.    8.  ]
 [ 9.1   0.4   0.5   1.8   0.07  7.   16.    0.99  3.21  0.69 12.5   8.  ]]


Finding all wines with quality number 5 and calculating the alverage of all alcohol, ph and sulphates based on the quality

In [27]:
quality3_col=column_names=="quality"
quality3_val=data[:,quality3_col].flatten()==5

all_wines_with_quality_equal_5=data[quality3_val]

#The index of alcohol=10
#The index of pH=8
#The index of Sulphates=9

print(f"Average Alcohol where its quality equals 5: {np.mean(all_wines_with_quality_equal_5[:,10]):.2f}\n")
print(f"Average pH where its quality equals 5: {np.mean(all_wines_with_quality_equal_5[:,9]):.2f}\n")
print(f"Average sulphates where its quality equals 5: {np.mean(all_wines_with_quality_equal_5[:,8]):.2f}\n")



Average Alcohol where its quality equals 5: 9.90

Average pH where its quality equals 5: 0.62

Average sulphates where its quality equals 5: 3.30



Answering the question, "The characteristics appear to distinguish higher-quality red wines from lower-quality red wines:

In [28]:
#locating the quality data
h_quality_wine=column_names=="quality"
alcohol_c=column_names=="alcohol"
volatile_acidity_c=column_names=="volatile acidity"

#finding the average quality wine
avg_quality=np.mean(data[:,h_quality_wine])

print(f"The average quality wine is {avg_quality:.2f}")

#finding the highest quality wine
highest_quality_wine=np.max(data[:,h_quality_wine])
print(f"The highest quality wine is {highest_quality_wine:.2f}")

#finding the lowest quality wine
lowest_quality_wine=np.min(data[:,h_quality_wine])
print(f"The lowest quality wine is {lowest_quality_wine:.2f}")


The average quality wine is 5.64
The highest quality wine is 8.00
The lowest quality wine is 3.00


Exmaining Alcohol content

In [29]:
#Filtering the range of quality wines below and beyond average

range_of_quality=data[:,h_quality_wine].flatten()
val_high_quality=range_of_quality>avg_quality
val_low_quality=range_of_quality<avg_quality

#wines that have high quality
wines_determined_with_high_quality=data[val_high_quality]

#wines that have low quality
wines_determined_with_low_quality=data[val_low_quality]

#determining whether high wine has a strong relationship with alcohol
print(f"Mean for all the alcohol before filtering: {np.mean(data[:,alcohol_c]):.2f}")

print(f"Mean for alcohol with high quality {np.mean(wines_determined_with_high_quality[:, alcohol_c]):.2f}")
print("\n")
print(f"Mean for alcohol with low quality: {np.mean(wines_determined_with_low_quality[:, alcohol_c]):.2f}")





Mean for all the alcohol before filtering: 10.42
Mean for alcohol with high quality 10.86


Mean for alcohol with low quality: 9.93


Alcohol Analysis

Key Finding                                                                         
Higher-quality wines have a higher average alcohol content (10.86%) than lower-quality wines (9.93%). The overall average alcohol content across all wines is 10.42%.

Interpretation

This suggests a positive association between alcohol content and wine quality. In this dataset, wines with higher quality scores tend to have higher alcohol content on average.

Limitation

This analysis identifies an association only. It does not demonstrate that increasing alcohol content causes a wine to have a higher quality rating. Other chemical properties may also contribute to wine quality.

Examining volatile acidity

In [30]:
#Displying mean of volatile acidity before filtering 
print(f"Mean of volatile acidity before filtering: {np.mean(data[:,volatile_acidity_c]):.2f}")

volatile_acidity_high_quality=data[val_high_quality]
volatile_acidity_low_quality=data[val_low_quality]

#Displaying mean of volatile acidity after filtering for high and low quality wines
print(f"Mean of volatile acidity with high quality: {np.mean(volatile_acidity_high_quality[:,volatile_acidity_c]):.2f}")
print(f"Mean of volatile acidity with low quality: {np.mean(volatile_acidity_low_quality[:,volatile_acidity_c]):.2f}")


Mean of volatile acidity before filtering: 0.53
Mean of volatile acidity with high quality: 0.47
Mean of volatile acidity with low quality: 0.59


Volatile Acidity Analysis                                                                         
Key Finding

Higher-quality wines have a lower average volatile acidity (0.47) than lower-quality wines (0.59). The overall average volatile acidity across all wines is 0.53.

Interpretation

This suggests a negative association between volatile acidity and wine quality. Wines with lower volatile acidity tend to receive higher quality ratings in this dataset.

Limitation

This analysis identifies an association only. It does not establish that reducing volatile acidity alone will improve wine quality, as multiple factors influence the final quality rating.

Examining sulphate

In [31]:
#Determining whether sulphates has a positive or negative association with wine quality
sulphates_col=column_names=="sulphates"
sulphates_high_quality=data[val_high_quality]
sulphates_low_quality=data[val_low_quality]
print(f"Mean of sulphates before filtering: {np.mean(data[:,sulphates_col]):.2f}")

print(f"Mean of sulphates with high quality: {np.mean(sulphates_high_quality[:,sulphates_col]):.2f}")
print(f"Mean of sulphates with low quality: {np.mean(sulphates_low_quality[:,sulphates_col]):.2f}")

Mean of sulphates before filtering: 0.66
Mean of sulphates with high quality: 0.69
Mean of sulphates with low quality: 0.62


Sulphates Analysis

Key Finding:                                                                     
Higher-quality wines have a higher average sulphates content (0.69) than lower-quality wines (0.62).

Interpretation:
This suggests a positive association between sulphates and wine quality.

Limitation:
The analysis identifies an association only and does not establish a causal relationship.

Examining Citric Acid

In [32]:
#determining whether citric acid has a positive or negative association with wine quality
citric_acid_col=column_names=="citric acid"
citric_acid_high_quality=data[val_high_quality]
citric_acid_low_quality=data[val_low_quality]

print(f"Mean of citric acid before filtering: {np.mean(data[:,citric_acid_col]):.2f}")
print(f"Mean of citric acid with high quality: {np.mean(citric_acid_high_quality[:,citric_acid_col]):.2f}")
print(f"Mean of citric acid with low quality: {np.mean(citric_acid_low_quality[:,citric_acid_col]):.2f}")   

Mean of citric acid before filtering: 0.27
Mean of citric acid with high quality: 0.30
Mean of citric acid with low quality: 0.24


Citric Acid Analysis                                                            
Key Finding                                                                     

Higher-quality wines have a higher average citric acid content (0.30) than lower-quality wines (0.24). The overall average citric acid content across all wines is 0.27.

Interpretation

This suggests a positive association between citric acid content and wine quality. Wines with higher quality scores tend to have slightly higher citric acid content on average.

Limitation

This analysis identifies an association only. It does not imply that increasing citric acid content alone will improve wine quality, as several chemical properties contribute to the final quality rating.

Examinig pH

In [33]:
#Determining whether pH has a positive or negative association with wine quality
pH_col=column_names=="pH"
pH_high_quality=data[val_high_quality]
pH_low_quality=data[val_low_quality]

print(f"Mean of pH before filtering: {np.mean(data[:,pH_col]):.2f}")
print(f"Mean of pH with high quality: {np.mean(pH_high_quality[:,pH_col]):.2f}")
print(f"Mean of pH with low quality: {np.mean(pH_low_quality[:, pH_col]):.2f}")

Mean of pH before filtering: 3.31
Mean of pH with high quality: 3.31
Mean of pH with low quality: 3.31


pH Analysis

Key Finding

Higher-quality wines and lower-quality wines both have an average pH of 3.31. The overall average pH across all wines is also 3.31.

Interpretation

This suggests little to no apparent association between pH and wine quality. Based on the average values, pH does not appear to distinguish higher-quality wines from lower-quality wines in this dataset.

Limitation

This conclusion is based on a comparison of average pH values only. While the averages are nearly identical, a more detailed statistical analysis (such as examining the distribution or performing hypothesis testing) could reveal subtler differences.

Examining Density

In [34]:
#Determining whether density has a positive or negative association with wine quality
density_col=column_names=="density"
density_high_quality=data[val_high_quality]
density_low_quality=data[val_low_quality]

print(f"Mean of density before filtering: {np.mean(data[:,density_col]):.2f}")
print(f"Mean of density with high quality: {np.mean(density_high_quality[:,density_col]):.2f}")
print(f"Mean of density with low quality: {np.mean(density_low_quality[:,density_col]):.2f}")

Mean of density before filtering: 1.00
Mean of density with high quality: 1.00
Mean of density with low quality: 1.00


Density Analysis

Key Finding

Higher-quality wines and lower-quality wines have nearly identical average density values (1.00). The overall average density across all wines is also 1.00.

Interpretation

This suggests little to no apparent association between density and wine quality. Based on the average values, density does not appear to distinguish higher-quality wines from lower-quality wines in this dataset.

Limitation

This conclusion is based on average density values only. Similar averages do not necessarily mean the full distributions are identical. Additional analyses, such as examining variability or statistical significance, could provide further insight.

## Summary Table

| Characteristic | Higher-Quality Wines | Lower-Quality Wines | Association |
|----------------|---------------------:|--------------------:|-------------|
| Alcohol | 10.86 | 9.93 | Positive |
| Volatile Acidity | 0.47 | 0.59 | Negative |
| Citric Acid | 0.30 | 0.24 | Positive |
| Sulphates | 0.69 | 0.62 | Positive |
| pH | 3.31 | 3.31 | Little to No Apparent Association |
| Density | 1.00 | 1.00 | Little to No Apparent Association |

## Overall Conclusion

This exploratory analysis investigated how selected chemical properties relate to wine quality using NumPy. The results indicate that higher-quality wines generally have higher average alcohol, sulphates, and citric acid contents, while exhibiting lower volatile acidity than lower-quality wines. In contrast, pH and density showed little to no apparent difference between the two quality groups based on their average values.

These findings suggest that alcohol, sulphates, citric acid, and volatile acidity are more strongly associated with wine quality than pH and density within this dataset. However, this analysis identifies associations only and does not establish causal relationships. Additional statistical analyses and machine learning techniques would be required to determine the predictive importance of these variables.

## Lessons Learned

Throughout this project, I strengthened my understanding of NumPy by applying array operations to a real-world dataset. I became more confident using Boolean masking, indexing, slicing, sorting, and statistical functions to answer practical data analysis questions.

I also learned the importance of distinguishing between association and causation when interpreting results, as well as presenting findings in a structured and professional manner.
